### PR TITLE
aerc: add oauth params

### PR DESCRIPTION
**Pull request by patwid**
Support oauth params for authentication mechanisms oauthbearer and xoauth2.

This is useful e.g. when adding a microsoft account as documented here: https://man.sr.ht/~rjarry/aerc/providers/microsoft.md

### CODE REVIEW

File: modules/programs/aerc-accounts.nix [hunk 1]
```diff
@@ -58,6 +69,25 @@ in {
...
+                scope = mkOption { type = nullOr (str); };
+              };
+            }));
+          default = null;
+          example =
+            literalExpression ''{ token_endpoint = "<token_endpoint>"; }'';
+          description = ''
+            Sets the oauth2 params if authentication mechanism oauthbearer or
+            xoauth2 is used.
+            See aerc-imap(5).
```
Comment (ncfavier) [inline modules/programs/aerc-accounts.nix:88]:
```suggestion
            See <citerefentry><refentrytitle>aerc-imap</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
```

File: modules/programs/aerc-accounts.nix [hunk 2]
```diff
@@ -58,6 +69,25 @@ in {
...
           '';
         };
+        oauth2Params = mkOption {
+          type = with types;
+            nullOr (attrsOf (submodule {
+              options = {
+                token_endpoint = mkOption { type = nullOr (str); };
+                client_id = mkOption { type = nullOr (str); };
+                client_secret = mkOption { type = nullOr (str); };
+                scope = mkOption { type = nullOr (str); };
```
[the remark this answers is not in the document]
Comment (patwid) [reply modules/programs/aerc-accounts.nix]: You are right, removed parentheses and specified a default value.

File: modules/programs/aerc-accounts.nix [hunk 3]
```diff
@@ -58,6 +69,25 @@ in {
...
+              };
+            }));
+          default = null;
+          example =
+            literalExpression ''{ token_endpoint = "<token_endpoint>"; }'';
```
[the remark this answers is not in the document]
Comment (patwid) [reply modules/programs/aerc-accounts.nix]: Removed `literalExpression` and also extra `attrsOf`.

File: modules/programs/aerc-accounts.nix [hunk 4]
```diff
@@ -69,25 +99,45 @@ in {
         if v != null && v != [ ] && v != "" then { ${k} = v; } else { };
       optPwCmd = k: p:
         optAttr "${k}-cred-cmd" (nullOrMap (builtins.concatStringsSep " ") p);
+      useOauth = auth: builtins.elem auth [ "oauthbearer" "xoauth2" ];
+      oauthParams = { auth, oauth2Params }:
+        if useOauth auth && oauth2Params != null && oauth2Params != { } then
+          "?" + builtins.concatStringsSep "&" lib.attrsets.mapAttrsToList
+          (k: v: k + "=" + v) oauth2Params
```
[the remark this answers is not in the document]
Comment (patwid) [reply modules/programs/aerc-accounts.nix:118]: Probably a good idea. How would we do that in nix?

[the remark this answers is not in the document]
Comment (ncfavier) [reply modules/programs/aerc-accounts.nix:118]: That's an open problem 😅. @RaitoBezarius wrote a urlencode function https://github.com/nix-community/home-manager/pull/3143, maybe you could coordinate.

[the remark this answers is not in the document]
Comment (patwid) [reply modules/programs/aerc-accounts.nix:118]: I have opened a pr to add a method `escapeQuery` for url encoding: https://github.com/NixOS/nixpkgs/pull/219235

[the remark this answers is not in the document]
Comment (patwid) [reply modules/programs/aerc-accounts.nix]: We now url-encode the values using the newly implemented function.

**Comment by patwid** [comment]
Would it make sense to define imap and smtp as submodules accepting auth and params for now? However, this would break current configurations..

**Comment by ncfavier** [comment]
The auth options would have different types, so probably not worth it.

We can merge this when https://nixpk.gs/pr-tracker.html?pr=219235 is all green.

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Pull request merged**